hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java

```diff
@@ -272,16 +274,49 @@ private B toBatchResult(F future) {
 
         private static final long serialVersionUID = 1L;
         private final BiFunctionEx<HazelcastInstance, InternalSerializationService, Reader<F, B, R>> readerSupplier;
+        private Map<Address, int[]> partitionAssignment = null;
+        private int[] partitionsToScan = null;
 
         LocalProcessorMetaSupplier(
                 @Nonnull BiFunctionEx<HazelcastInstance, InternalSerializationService, Reader<F, B, R>> readerSupplier
         ) {
             this.readerSupplier = readerSupplier;
         }
 
+        LocalProcessorMetaSupplier(
+                @Nonnull BiFunctionEx<HazelcastInstance, InternalSerializationService, Reader<F, B, R>> readerSupplier,
+                int[] partitionsToScan
+        ) {
+            this.readerSupplier = readerSupplier;
+            this.partitionsToScan = partitionsToScan;
+        }
+
         @Override @Nonnull
         public Function<Address, ProcessorSupplier> get(@Nonnull List<Address> addresses) {
-            return address -> new LocalProcessorSupplier<>(readerSupplier);
+            if (partitionsToScan == null) {
+                return address -> new LocalProcessorSupplier<>(readerSupplier);
+            } else {
+                return address -> {
+                    int[] partitions = partitionAssignment.get(address);
+                    Arrays.sort(partitions);
+                    List<Integer> partitionsToScanList = new ArrayList<>();
+                    for (int pId : partitionsToScan) {
+                        if (Arrays.binarySearch(partitions, pId) > 0) {
```

Comment:
you could use something similar to merge-join if `partitionsToScan` was sorted. Is it?